packages/caravan-fees/src/rbf.ts

```diff
@@ -0,0 +1,530 @@
+    dustThreshold,
+    network,
+    scriptType,
+    requiredSigners,
+    totalSigners,
+  });
+
+  // Step 4: Add the cancellation output and delete all previous outputs
+
+  newTxTemplate.addOutput(
+    new BtcTxOutputTemplate({
+      address: cancelAddress,
+      amountSats: "0", // Temporary amount, will be adjusted later
+      locked: false,
+    }),
+  );
+
+  // Step 5: Add inputs from the original transaction
+  const originalInputTemplates = txAnalyzer.getInputTemplates();
+  const addedInputs = new Set();
+
+  if (reuseAllInputs) {
+    // Add all original inputs
+    originalInputTemplates.forEach((template) => {
+      const input = availableInputs.find(
+        (utxo) => utxo.txid === template.txid && utxo.vout === template.vout,
+      );
+      if (input) {
+        newTxTemplate.addInput(BtcTxInputTemplate.fromUTXO(input));
+        addedInputs.add(`${input.txid}:${input.vout}`);
+      }
+    });
+  } else {
+    // Add at least one input from the original transaction
+
+    // TO DO (MRIGESH)
+    // Add coin selection algorithm to add best suited input to start with ...
+    const originalInput = availableInputs.find((utxo) =>
+      originalInputTemplates.some(
+        (template) =>
+          template.txid === utxo.txid && template.vout === template.vout,
+      ),
+    );
+    if (!originalInput) {
+      throw new Error(
+        "None of the original transaction inputs found in available UTXOs.",
+      );
+    }
+    newTxTemplate.addInput(BtcTxInputTemplate.fromUTXO(originalInput));
+    addedInputs.add(`${originalInput.txid}:${originalInput.vout}`);
+  }
+
+  // Step 6: Add more inputs if necessary to meet fee requirements
+
+  // We continue adding inputs until both the fee rate is satisfied and
+  // the absolute fee meets the minimum RBF requirement
+  for (const utxo of availableInputs) {
+    if (
+      newTxTemplate.feeRateSatisfied &&
+      new BigNumber(newTxTemplate.currentFee).gte(
+        BigNumber.max(newTxTemplate.targetFeesToPay, originalTxFee),
+      )
+    ) {
+      // Stop adding inputs when both conditions are met:
+      // 1. The fee rate satisfies the target rate
+      // 2. The current fee is greater than or equal to the maximum of:
+      //    a) The target fee based on the desired rate by user
+      //    b) The minimum fee required for a valid RBF replacement >= original fees
+      break;
+    }
+    // Skip if this UTXO is already added
+    const utxoKey = `${utxo.txid}:${utxo.vout}`;
+    if (addedInputs.has(utxoKey)) {
+      continue;
+    }
+    newTxTemplate.addInput(BtcTxInputTemplate.fromUTXO(utxo));
+    addedInputs.add(utxoKey);
+  }
+
+  // Step 7: Calculate and set the cancellation output amount
+
+  const totalInputAmount = new BigNumber(newTxTemplate.getTotalInputAmount());
+  // Ensure we're using the higher of the target fee and the original fee
+  const fee = BigNumber.max(newTxTemplate.targetFeesToPay, originalTxFee);
```

Comment:
if you validate this at the top, then this isn't necessary.

Reply:
No it's still required ...

Taking an example from the fixtures to understand why we need this , especially for cancellation case where our tx size decreases ...


>   Original Transaction:
      - Type: P2WSH (Pay-to-Witness-Script-Hash)
      - Inputs: 2
      - Outputs: 3
      - Fee: 20,880 satoshis
      - Size: 348 vBytes
      - Fee Rate: 60 sat/vB (20,880 / 348)

      RBF Cancellation Scenario:
      1. Target Fee Rate: 80 sat/vB (increased from original 60 sat/vB)
      2. Cancellation tx typically has fewer outputs (usually just one to the cancellation address)
      3. In this case, the new tx size decreases to 159 vBytes due to having only one output

      RBF Fee Calculation:
      - Minimum required fee: 20,880 satoshis (must be >= original fee to satisfy RBF rule #3)
      - Actual fee rate achieved: 131.32 sat/vB (20,880 / 159)

      Key Observation:
      Even though a fee of 12,720 satoshis (159 * 80) would achieve the desired 80 sat/vB rate,
      we must pay the original fee of 20,880 satoshis to satisfy RBF rules. This results in a
      higher effective fee rate than initially targeted.

>  This example illustrates how RBF cancellations can sometimes lead to paying more in fees
      than strictly necessary for the desired fee rate, due to the requirement to pay at least
      the original transaction's fee.
      
      


      So here `targetFeesToPay` would suggest us to pay `12,720 ` which is correct but not RBF wise